src/OpenTelemetry.Exporter.Geneva/Metrics/GenevaMetricExporter.cs

```diff
@@ -484,6 +486,384 @@ internal static void DisableOpenTelemetrySdkMetricNameValidation()
+
+            // Get a placeholder to add the payloadType length
+            payloadTypeStartIndex = bufferIndex;
+            bufferIndex += 2;
+
+            // Get a placeholder to add dimensions count later
+            var bufferIndexForDimensionsCount = bufferIndex;
+            bufferIndex += 2;
+
+            ushort dimensionsWritten = 0;
+
+            // Serialize PrepopulatedDimensions keys
+            for (ushort i = 0; i < this.prepopulatedDimensionsCount; i++)
+            {
+                MetricSerializer.SerializeEncodedString(this.buffer, ref bufferIndex, this.serializedPrepopulatedDimensionsKeys[i]);
+            }
+
+            if (this.prepopulatedDimensionsCount > 0)
+            {
+                dimensionsWritten += this.prepopulatedDimensionsCount;
+            }
+
+            // Serialize MetricPoint Dimension keys
+            foreach (var tag in tags)
+            {
+                if (tag.Key.Length > MaxDimensionNameSize)
+                {
+                    // TODO: Data Validation
+                }
+
+                MetricSerializer.SerializeString(this.buffer, ref bufferIndex, tag.Key);
+            }
+
+            dimensionsWritten += (ushort)tags.Count;
+
+            // Serialize PrepopulatedDimensions values
+            for (ushort i = 0; i < this.prepopulatedDimensionsCount; i++)
+            {
+                MetricSerializer.SerializeEncodedString(this.buffer, ref bufferIndex, this.serializedPrepopulatedDimensionsValues[i]);
+            }
+
+            // Serialize MetricPoint Dimension values
+            foreach (var tag in tags)
+            {
+                var dimensionValue = Convert.ToString(tag.Value, CultureInfo.InvariantCulture);
+                if (dimensionValue.Length > MaxDimensionValueSize)
+                {
+                    // TODO: Data Validation
+                }
+
+                MetricSerializer.SerializeString(this.buffer, ref bufferIndex, dimensionValue);
+            }
+
+            // Backfill the number of dimensions written
+            MetricSerializer.SerializeUInt16(this.buffer, ref bufferIndexForDimensionsCount, dimensionsWritten);
+
+            payloadTypeLength = (ushort)(bufferIndex - payloadTypeStartIndex - 2);
+            MetricSerializer.SerializeUInt16(this.buffer, ref payloadTypeStartIndex, payloadTypeLength);
+
+            #endregion
+
+            #region Serialize exemplars
+
+            if (exemplars.Length > 0)
+            {
+                MetricSerializer.SerializeByte(this.buffer, ref bufferIndex, (byte)PayloadType.Exemplars);
+
+                // Get a placeholder to add the payloadType length
+                payloadTypeStartIndex = bufferIndex;
+                bufferIndex += 2;
+
+                MetricSerializer.SerializeByte(this.buffer, ref bufferIndex, 0); // version
+
+                var exemplarsCount = 0;
+                foreach (var exemplar in exemplars)
```

Comment:
lets add a TODO to avoid this dual iteration, and instead backfill the count

Reply:
I don't think we can avoid it. The count field doesn't have a predefined size. It could be anywhere from 1 to 9 bytes. So we need to know the count beforehand.